### ISSUE TITLE
loss.backward() can't update model parameters

### ISSUE THREAD

**Issue by Wuyxin** [pytorch/pytorch#40983]
## ❓ Questions and Help

Here's a simple model with two self-designed pooling layers(self.node_pooling, self.edge_pooling), the problem is that the layers' **parameters remain unchanged** during training. Further, I check the **gradients** of some variables (e.g. self.model.readout, x, edge_attr) after loss.backward,  and all of these turn out to be **None**. 

```
......
optimizer = torch.optim.Adam([
                {'params': self.node_pool.parameters(), 'lr':lr},
                {'params': self.edge_pool.parameters(), 'lr':lr}
            ])
for i in range(1, epoch + 1):
            optimizer.zero_grad()
            x, edge_index, edge_attr, batch, perm, self.node_score = self.node_pool(subgraph.x, 
                                                                                    subgraph.edge_index, 
                                                                                    subgraph.edge_attr
                                                                                    )
            x, edge_index, edge_attr, indices, self.edge_score = self.edge_pool(x ,edge_index, edge_attr)
            
            graph = Graph(x = x, edge_index = edge_index, edge_attr = edge_attr)   # class : torch_geometric.data.Data
            graph.to(device)             # cuda
            self.model(graph, **kwargs)  # model forwards, self.model is a trained GNN
            loss = -self.model.readout[new_target_idx, self.graph.y[target]]       # self.model.readout : [num_nodes, num_classes]
            k  = Variable(x.cuda(), requires_grad=True)
            k.retain_grad()
            loss.backward()
            print(k.grad, k.requires_grad) # None True
            optimizer.step()
......
```

I don't know if the way I use autograd to check gradients is wrong or the all-None gradients prevent loss to pass backward and thus prevent the parameters of the layers to update.  

I will be really appreciate if somebody could tell me what's wrong

**Comment by MANISH007700** [comment]
After the optimizer.step() , new weights and bias are made , so you have to clear the old bias and weight  and hence you should add optimizer.zer0_grad() after optimizer.step() , so that new weights can be updated !

Cheers !

**Comment by albanD** [comment]
Hi,

We use github issues only for bugs or feature requests.
Please use the forum to ask questions: https://discuss.pytorch.org/

For your particular question,
`k` is not used to compute the loss, so it is expected that it has no gradient.
For things like `model.readout`, you seem to only use a subset of it, so I guess that the original Tensor's .grad field is not populated indeed.

Don't hesitate to open a new topic on the forum with the code sample showing how you try to access these gradients.
